calico-vpp-agent/connectivity/ipsec.go

```diff
@@ -111,211 +133,178 @@ func (p *IpsecProvider) RescanState() {
 	if err != nil {
-		p.errorCleanup(tunnel, profile)
 		return errors.Wrapf(err, "error configuring IPsec tunnel %s", tunnel.String())
 	}
 
-	err = p.vpp.SetIKEv2LocalIDAddress(profile, tunnel.Src)
+	err = p.vpp.SetIKEv2LocalIDAddress(tunnel.Profile(), tunnel.Src)
 	if err != nil {
-		p.errorCleanup(tunnel, profile)
 		return errors.Wrapf(err, "error configuring IPsec tunnel %s", tunnel.String())
 	}
 
-	err = p.vpp.SetIKEv2RemoteIDAddress(profile, tunnel.Dst)
+	err = p.vpp.SetIKEv2RemoteIDAddress(tunnel.Profile(), tunnel.Dst)
 	if err != nil {
-		p.errorCleanup(tunnel, profile)
 		return errors.Wrapf(err, "error configuring IPsec tunnel %s", tunnel.String())
 	}
 
-	err = p.vpp.SetIKEv2PermissiveTrafficSelectors(profile)
+	err = p.vpp.SetIKEv2PermissiveTrafficSelectors(tunnel.Profile())
 	if err != nil {
-		p.errorCleanup(tunnel, profile)
 		return errors.Wrapf(err, "error configuring IPsec tunnel %s", tunnel.String())
 	}
 
 	// Compare addresses lexicographically to select an initiator
-	if bytes.Compare(tunnel.Src.To4(), tunnel.Dst.To4()) > 0 {
+	if tunnel.IsInitiator() {
 		p.log.Infof("connectivity(add) IKE Set responder=%s", tunnel.String())
-		err = p.vpp.SetIKEv2Responder(profile, config.DataInterfaceSwIfIndex, tunnel.Dst)
+		err = p.vpp.SetIKEv2Responder(tunnel.Profile(), config.DataInterfaceSwIfIndex, tunnel.Dst)
 		if err != nil {
-			p.errorCleanup(tunnel, profile)
 			return errors.Wrapf(err, "error configuring IPsec tunnel %s", tunnel.String())
 		}
 
-		err = p.vpp.SetIKEv2DefaultTransforms(profile)
+		err = p.vpp.SetIKEv2DefaultTransforms(tunnel.Profile())
 		if err != nil {
-			p.errorCleanup(tunnel, profile)
 			return errors.Wrapf(err, "error configuring IPsec tunnel %s", tunnel.String())
 		}
 
-		err = p.vpp.IKEv2Initiate(profile)
+		err = p.vpp.IKEv2Initiate(tunnel.Profile())
 		if err != nil {
-			p.errorCleanup(tunnel, profile)
 			return errors.Wrapf(err, "error configuring IPsec tunnel %s", tunnel.String())
 		}
 	}
+	p.log.Infof("connectivity(add) IPsec tunnel=%s", tunnel.String())
 
 	// Wait for IPsec connection to be established to bring tunnel up
-	go p.waitForIPsecSA(profile, tunnel)
+	tunnel.cancel = p.waitForIPsecSA(*tunnel)
 
 	return nil
 }
 
-func (p *IpsecProvider) waitForIPsecSA(profile string, tunnel *types.IPIPTunnel) {
-	for {
-		time.Sleep(time.Second)
-		iface, err := p.vpp.GetInterfaceDetails(tunnel.SwIfIndex)
-		if err != nil {
-			p.log.WithError(err).Errorf("Cannot get IPIP tunnel %s status", tunnel.String())
-			return
-		}
-		if iface.IsUp {
-			p.log.Infof("connectivity(add) tunnel now up Profile=%s", profile)
-			return
-		}
-
-		if bytes.Compare(tunnel.Src.To4(), tunnel.Dst.To4()) > 0 {
-			p.log.Warnf("IPIP tunnel still down, re-trying initiate IKE for IPsec tunnel=%s", tunnel.String())
-			err = p.vpp.IKEv2Initiate(profile)
+func (p *IpsecProvider) waitForIPsecSA(tunnel IpsecTunnel) func() {
+	alive := true
+	go (func() {
+		for alive {
+			time.Sleep(time.Second)
```

Comment:
We sleep after checking for alive, so the tunnel may have been deleted before the next API calls.
It might actually be cleaner to do this with a `close` channel inside the IpsecTunnel object instead of a cancel func, then the wait function can be written with a select on that channel and on a ticker, and cancelling the tunnel retries is as simple as closing the channel.